consensus/hotstuff/eventhandler/event_handler_test.go

```diff
@@ -388,29 +307,160 @@ func (es *EventHandlerSuite) SetupTest() {
+	err := es.eventhandler.OnReceiveProposal(proposal)
+	require.NoError(es.T(), err)
+	es.voteAggregator.AssertNotCalled(es.T(), "AddBlock", proposal)
 }
 
-// a QC for current view triggered view change
-func (es *EventHandlerSuite) TestQCBuiltViewChanged() {
+// TestOnReceiveProposal_QCOlderThanCurView tests scenario: received a valid proposal with QC that has older view,
+// the proposal's QC shouldn't trigger view change.
+func (es *EventHandlerSuite) TestOnReceiveProposal_QCOlderThanCurView() {
+	proposal := createProposal(es.initView-1, es.initView-2)
+	es.voteAggregator.On("AddBlock", proposal).Return(nil).Once()
+
+	// should not trigger view change
+	err := es.eventhandler.OnReceiveProposal(proposal)
+	require.NoError(es.T(), err)
+	require.Equal(es.T(), es.endView, es.paceMaker.CurView(), "incorrect view change")
+	es.voteAggregator.AssertCalled(es.T(), "AddBlock", proposal)
+}
+
+// TestOnReceiveProposal_TCOlderThanCurView tests scenario: received a valid proposal with QC and TC that has older view,
+// the proposal's QC shouldn't trigger view change.
+func (es *EventHandlerSuite) TestOnReceiveProposal_TCOlderThanCurView() {
+	proposal := createProposal(es.initView-1, es.initView-3)
+	proposal.LastViewTC = helper.MakeTC(helper.WithTCView(proposal.Block.View-1), helper.WithTCNewestQC(proposal.Block.QC))
+	es.voteAggregator.On("AddBlock", proposal).Return(nil).Once()
+
+	// should not trigger view change
+	err := es.eventhandler.OnReceiveProposal(proposal)
+	require.NoError(es.T(), err)
+	require.Equal(es.T(), es.endView, es.paceMaker.CurView(), "incorrect view change")
+	es.voteAggregator.AssertCalled(es.T(), "AddBlock", proposal)
+}
+
+// TestOnReceiveProposal_NoVote tests scenario: received a valid proposal for cur view, but not a safe node to vote, and I'm the next leader
+// should not vote.
+func (es *EventHandlerSuite) TestOnReceiveProposal_NoVote() {
+	proposal := createProposal(es.initView, es.initView-1)
+	es.voteAggregator.On("AddBlock", proposal).Return(nil).Once()
+
+	// I'm the next leader
+	es.committee.leaders[es.initView+1] = struct{}{}
+	// no vote for this proposal
+	err := es.eventhandler.OnReceiveProposal(proposal)
+	require.NoError(es.T(), err)
+	require.Equal(es.T(), es.endView, es.paceMaker.CurView(), "incorrect view change")
+	es.voteAggregator.AssertCalled(es.T(), "AddBlock", proposal)
+}
+
+// TestOnReceiveProposal_NoVote_ParentProposalNotFound tests scenario: received a valid proposal for cur view, no parent for this proposal found
+// should not vote.
+func (es *EventHandlerSuite) TestOnReceiveProposal_NoVote_ParentProposalNotFound() {
+	proposal := createProposal(es.initView, es.initView-1)
+	es.voteAggregator.On("AddBlock", proposal).Return(nil).Once()
+
+	// remove parent from known proposals
+	delete(es.forks.proposals, proposal.Block.QC.BlockID)
+
+	// no vote for this proposal, no parent found
+	err := es.eventhandler.OnReceiveProposal(proposal)
+	require.NoError(es.T(), err)
```

Comment:
Mentioned in a separate comment - this case it not expected so I feel like it should return an error, rather than log.